src/navigation/failure_identification/failure_identification.py

```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env python3
+import rospy
+import message_filters
+from mrover.msg import MotorsStatus
+from geometry_msgs.msg import Twist
+from nav_msgs.msg import Odometry
+from smach_msgs.msg import SmachContainerStatus
+from std_msgs.msg import Bool
+import pandas as pd
+from pandas import DataFrame
+from watchdog import WatchDog
+import numpy as np
+import os
+from pathlib import Path
+from util.ros_utils import get_rosparam
+
+DATAFRAME_MAX_SIZE = 50
+
+
+class FailureIdentifier:
+    """
+    Used to identify if the rover is stuck. Owns a data frame that is updated with the latest data from the rover
+    and then used to determine if the rover is stuck
+    """
+
+    stuck_publisher: rospy.Publisher
+    _df: DataFrame
+    watchdog: WatchDog
+    actively_collecting: bool
+    cur_cmd: Twist
+    cur_stuck: bool
+    path_name: Path
+
+    def __init__(self):
+        nav_status_sub = message_filters.Subscriber("smach/container_status", SmachContainerStatus)
+        cmd_vel_sub = rospy.Subscriber("cmd_vel", Twist, self.cmd_vel_update)
+        drive_status_sub = message_filters.Subscriber("drive_status", MotorsStatus)
+        odometry_sub = message_filters.Subscriber("global_ekf/odometry", Odometry)
+        stuck_button_sub = rospy.Subscriber("/rover_stuck", Bool, self.stuck_button_update)
+
+        ts = message_filters.ApproximateTimeSynchronizer(
+            [nav_status_sub, drive_status_sub, odometry_sub], 10, 1.0, allow_headerless=True
+        )
+        ts.registerCallback(self.update)
+
+        self.stuck_publisher = rospy.Publisher("/nav_stuck", Bool, queue_size=1)
+        position_variables = ["x", "y", "z"]
+        rotation_variables = [f"rot_{comp}" for comp in ["x", "y", "z", "w"]]
+        velocity_variables = ["linear_velocity", "angular_velocity"]
+        wheel_effort_variables = [f"wheel_{wheel_num}_effort" for wheel_num in range(6)]
+        wheel_velocity_variables = [f"wheel_{wheel_num}_velocity" for wheel_num in range(6)]
+        command_variables = ["cmd_vel_x", "cmd_vel_twist"]
+        self.data_collecting_mode = True
+        self.actively_collecting = True
+        self.cur_cmd = Twist()
+        self.cur_stuck = False
+        self.cols = (
+            ["time", "stuck"]
+            + position_variables
+            + rotation_variables
+            + velocity_variables
+            + wheel_effort_variables
+            + wheel_velocity_variables
+            + command_variables
+        )
+        print(self.cols)
+        self._df = pd.DataFrame(columns=self.cols)
+        self.left_pointer = 0
+        self.right_pointer = 0
+        self.row_counter = 0
+        self.watchdog = WatchDog(self)
+        self.path_name = None  # type: ignore
+
+    def write_to_csv(self):
+        """
+        Writes the data frame to a csv file marked with timestamp
+        """
+        home = Path.home()
+        path = home / "catkin_ws/src/mrover/src/failure_data"
```

Comment:
remove the second src, I think this should be a folder in the top level mrover directory, not in src